### PR TITLE
Fix GitHub’s own "Edit readme" link on tag pages

### LINKED ISSUE #2997: GitHub’s own "Edit readme" link points to 404 on tag pages
1. Visit a repo you own
2. Open the latest tag (like https://github.com/sindresorhus/refined-github/tree/20.4.9)
2. Click edit button on top right of readme
3. See 404 at `/$USER/$REPO/edit/$TAG/readme.md`

This button is added by GitHub, it's their own bug.

---

On tag pages, that link should edit the readme on the default branch.

The `edit-readme` feature should be extended to also fix this GitHub bug.

Previously discussed in https://github.com/sindresorhus/refined-github/issues/2982

### PR DESCRIPTION
**Pull request by yakov116**
By extending `edit-readme`

Fixes #2997 

Test on any repo you have access to.

I tested on https://github.com/yakov116/TestR/tree/1.6
<!--

Thanks for contributing! 🍄

1. LINKED ISSUES:
   Does this PR close/fix an existing issue? Write something like `Closes #10`

2. TEST URLS:
   Add some test URLs

3. SCREENSHOT:
   Add a screenshot here if your PR makes visual changes

### CODE REVIEW

File: source/features/edit-readme.tsx [hunk 1]
```diff
@@ -24,6 +19,17 @@ async function init(): Promise<void | false> {
 		pathnameParts[4] = await getDefaultBranch(); // Replaces /${tag|commit}/
 	}
 
+	// The button already exists on repos you can push to.
+	const buttonExists = select<HTMLAnchorElement>('a[aria-label="Edit this file"]');
+	if (buttonExists) {
+		// See https://github.com/sindresorhus/refined-github/issues/2997: Fixes GitHub’s own "Edit readme" link on tag pages
+		if (isPermalink) {
```
Comment (fregante) [inline source/features/edit-readme.tsx:26]:
```suggestion
		if (isPermalink) {
			// GitHub has a broken link in this case #2997
```

File: source/features/edit-readme.tsx [hunk 2]
```diff
@@ -24,6 +19,17 @@ async function init(): Promise<void | false> {
 		pathnameParts[4] = await getDefaultBranch(); // Replaces /${tag|commit}/
 	}
 
+	// The button already exists on repos you can push to.
+	const buttonExists = select<HTMLAnchorElement>('a[aria-label="Edit this file"]');
+	if (buttonExists) {
```
Comment (fregante) [inline source/features/edit-readme.tsx:24]:
```suggestion
	if (existingButton) {
```

File: source/features/edit-readme.tsx [hunk 3]
```diff
@@ -24,6 +19,17 @@ async function init(): Promise<void | false> {
...
+		// See https://github.com/sindresorhus/refined-github/issues/2997: Fixes GitHub’s own "Edit readme" link on tag pages
+		if (isPermalink) {
+			buttonExists.href = pathnameParts.join('/');
```
Comment (fregante) [inline source/features/edit-readme.tsx:27]:
```suggestion
			existingButton.href = pathnameParts.join('/');
```

**Comment by fregante** [comment]
In a separate PR, you could also add support for branches with slashes in the name in this feature and any feature that makes use of the also-affected `getReference`

File: readme.md [hunk 4]
```diff
@@ -144,7 +144,7 @@ Thanks for contributing! 🦋🙌
 - [](# "download-folder-button") [Adds a button to a download button entire folders.](https://user-images.githubusercontent.com/1402241/35044451-fd3e2326-fbc2-11e7-82e1-61ec7bee612b.png) *Uses [download-directory.github.io](https://download-directory.github.io)*
 - [](# "toggle-files-button") [Adds a button to toggle the repo file list.](https://user-images.githubusercontent.com/1402241/35480123-68b9af1a-043a-11e8-8934-3ead3cff8328.gif)
 - [](# "edit-files-faster") [Adds a button to edit files from the repo file list.](https://user-images.githubusercontent.com/1402241/56370462-d51cde00-622d-11e9-8cd3-8a173bd3dc08.png)
```
Comment (fregante) [inline readme.md:146]: Yo

**Comment by yakov116** [comment]
> In a separate PR, you could also add support for branches with slashes in the name in this feature (and any feature that makes use of `pathnameParts[4]` and the also-affected `getReference`)
> 
> `fix-view-file-link-in-pr` and `link-to-file-in-file-history` both have solutions for it, but I think they're not easily copy-able.

I just tested it and it looks like its working fine 
https://github.com/yakov116/TestR/tree/1.8

**Comment by fregante** [comment]
`1.8` doesn’t contain slashes (`/`)

Also, I was kinda wrong, this doesn’t affect branches with slashes BUT it does affect tags with slashes (if that’s possible)

**Comment by yakov116** [comment]
> `1.8` doesn’t contain slashes (`/`)
> 
> Also, I was kinda wrong, this doesn’t affect branches with slashes BUT it does affect tags with slashes (if that’s possible)

Not possible. Via website will lead to a 500 and via api will not create a tag

**Comment by fregante** [comment]
Possible via git: https://github.com/fregante/slash-tag-test/tree/a/b

**Comment by yakov116** [comment]
> Possible via git: [fregante/slash-tag-test:b@`a`](https://github.com/fregante/slash-tag-test/tree/a/b) [](/fregante/slash-tag-test/tree/HEAD@{2020-04-14T13:49:42Z}/b)

And I thought i tried everthing :)

I think that its very very rare. Think its worth dealing with it?

**Comment by fregante** [comment]
Not a big deal to support it but maybe there's an easy way to get the reference without using the URL.

For example `getReference` should be changed to:
```js
// Tested on the Code tab
return select('#branch-select-menu summary .css-truncate-target').textContent
```

**Comment by yakov116** [comment]
Let's merge this I will not have time until next week. I will open a new PR.

**Comment by fregante** [comment]
> In a separate PR, you could also add support for branches with slashes

Yeah that was the intention ⬆️

**Pull request merged**